File: code/lecture04/c/quickselect.c
#include <assert.h>
#include <limits.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void swap_int(int *a, int *b) {
    int t = *a; *a = *b; *b = t;
}

/* Partitions [lo, hi) into < pivot, == pivot, > pivot. */
static void partition3(int *a, size_t lo, size_t hi, int pivot,
                       size_t *lt_out, size_t *gt_out) {
    size_t lt = lo, scan = lo, gt = hi;
    while (scan < gt) {
        if (a[scan] < pivot) swap_int(&a[lt++], &a[scan++]);
        else if (a[scan] > pivot) swap_int(&a[scan], &a[--gt]);
        else ++scan;
    }
    *lt_out = lt; *gt_out = gt; /* equal range is [lt, gt) */
}

/*
 * RandomizedSelect: selects zero-based relative rank from a[0..n).
 * The array is modified; the random pivot policy is internal to the algorithm.
 * Returns false for NULL pointers, n == 0, or rank >= n.
 * RNG policy belongs to the caller: seed rand() once (not per call).
 */
bool quickselect_int(int *a, size_t n, size_t rank, int *result) {
    if (a == NULL || result == NULL || n == 0 || rank >= n) return false;
    size_t lo = 0, hi = n;
    for (;;) {
        if (hi - lo == 1) { *result = a[lo]; return true; }
        size_t pivot_index = lo + (size_t)rand() % (hi - lo);
        int pivot = a[pivot_index];
        size_t lt, gt;
        partition3(a, lo, hi, pivot, &lt, &gt);
        if (rank < lt) hi = lt;
        else if (rank >= gt) lo = gt;
        else { *result = pivot; return true; }
    }
}

static int compare_int(const void *pa, const void *pb) {
    int a = *(const int *)pa, b = *(const int *)pb;
    return (a > b) - (a < b);
}

static void verify_all_ranks(const int *input, size_t n, unsigned seed) {
    int *oracle = malloc(n * sizeof *oracle);
    int *work = malloc(n * sizeof *work);
    assert(oracle != NULL && work != NULL);
    memcpy(oracle, input, n * sizeof *oracle);
    qsort(oracle, n, sizeof *oracle, compare_int);
    for (size_t rank = 0; rank < n; ++rank) {
        memcpy(work, input, n * sizeof *work);
        srand(seed + (unsigned)rank);
        int result = 0;
        assert(quickselect_int(work, n, rank, &result));
        assert(result == oracle[rank]);
    }
    free(work);
    free(oracle);
}

int main(void) {
    const int one[] = {7};
    const int two[] = {2, 1};
    const int ascending[] = {1, 2, 3, 4, 5, 6};
    const int descending[] = {6, 5, 4, 3, 2, 1};
    const int equal[] = {4, 4, 4, 4, 4, 4};
    const int duplicates[] = {4, 2, 4, 1, 4, 3, 2, 4, 1};
    const int not_five[] = {31, 8, 48, 73, 11, 3, 20};
    const int extremes[] = {INT_MAX, 0, INT_MIN, -1, 1, INT_MAX};
    const struct { const int *a; size_t n; } cases[] = {
        {one, 1}, {two, 2}, {ascending, 6}, {descending, 6},
        {equal, 6}, {duplicates, 9}, {not_five, 7}, {extremes, 6}
    };
    for (unsigned seed = 1; seed <= 32; ++seed) {
        for (size_t c = 0; c < sizeof cases / sizeof cases[0]; ++c) {
            verify_all_ranks(cases[c].a, cases[c].n, seed * 101U);
        }
    }

    int out;
    int valid[] = {1, 2};
    assert(!quickselect_int(valid, 2, 2, &out));
    assert(!quickselect_int(valid, 0, 0, &out));
    assert(!quickselect_int(valid, 2, 0, NULL));
    assert(!quickselect_int(NULL, 0, 0, &out));
    puts("RandomizedSelect C tests passed");
    return 0;
}

File: code/lecture04/java/Quickselect.java
import java.util.Arrays;
import java.util.Objects;
import java.util.Random;

public final class Quickselect {
    private Quickselect() {}

    /**
     * RandomizedSelect for a zero-based relative rank in a[0..a.length).
     * Modifies the array; the caller supplies the RNG policy.
     */
    public static int select(int[] a, int rank, Random random) {
        Objects.requireNonNull(a, "array");
        Objects.requireNonNull(random, "random");
        if (rank < 0 || rank >= a.length) {
            throw new IllegalArgumentException("rank outside [0, n)");
        }
        int lo = 0, hi = a.length;
        while (true) {
            if (hi - lo == 1) return a[lo];
            int pivot = a[lo + random.nextInt(hi - lo)];
            int[] equal = partition3(a, lo, hi, pivot);
            if (rank < equal[0]) hi = equal[0];
            else if (rank >= equal[1]) lo = equal[1];
            else return pivot;
        }
    }

    /** Returns [lt, gt), the interval equal to pivot. */
    private static int[] partition3(int[] a, int lo, int hi, int pivot) {
        int lt = lo, scan = lo, gt = hi;
        while (scan < gt) {
            if (a[scan] < pivot) swap(a, lt++, scan++);
            else if (a[scan] > pivot) swap(a, scan, --gt);
            else scan++;
        }
        return new int[] {lt, gt};
    }

    private static void swap(int[] a, int i, int j) {
        int t = a[i]; a[i] = a[j]; a[j] = t;
    }

    private static void require(boolean condition, String message) {
        if (!condition) throw new AssertionError(message);
    }

    private static void verifyAllRanks(int[] input, long seed) {
        int[] oracle = input.clone();
        Arrays.sort(oracle);
        for (int rank = 0; rank < input.length; rank++) {
            int actual = select(input.clone(), rank, new Random(seed + rank));
            require(actual == oracle[rank],
                    "rank " + rank + ": " + actual + " != " + oracle[rank]);
        }
    }

    private static void expectIllegalRank(int[] input, int rank) {
        try {
            select(input, rank, new Random(1));
            throw new AssertionError("invalid rank accepted: " + rank);
        } catch (IllegalArgumentException expected) {
            require("rank outside [0, n)".equals(expected.getMessage()),
                    "unexpected invalid-rank message");
        }
    }

    public static void main(String[] args) {
        int[][] cases = {
            {7},
            {2, 1},
            {1, 2, 3, 4, 5, 6},
            {6, 5, 4, 3, 2, 1},
            {4, 4, 4, 4, 4, 4},
            {4, 2, 4, 1, 4, 3, 2, 4, 1},
            {31, 8, 48, 73, 11, 3, 20},
            {Integer.MAX_VALUE, 0, Integer.MIN_VALUE, -1, 1, Integer.MAX_VALUE}
        };
        for (long seed = 1; seed <= 32; seed++) {
            for (int[] input : cases) verifyAllRanks(input, seed * 101);
        }

        expectIllegalRank(new int[] {1, 2}, -1);
        expectIllegalRank(new int[] {1, 2}, 2);
        expectIllegalRank(new int[0], 0);
        try {
            select(null, 0, new Random(1));
            throw new AssertionError("null array accepted");
        } catch (NullPointerException expected) {
            require("array".equals(expected.getMessage()), "unexpected null-array message");
        }
        try {
            select(new int[] {1}, 0, null);
            throw new AssertionError("null RNG accepted");
        } catch (NullPointerException expected) {
            require("random".equals(expected.getMessage()), "unexpected null-RNG message");
        }
        System.out.println("RandomizedSelect Java tests passed");
    }
}
